### PR TITLE
Fundamentals of container security

### LINKED ISSUE #2216: Fundamentals of container security
### Brief Summary

The purpose of this article is to highlight the security issues and how developers and other key players can mitigate them during the deployment and maintenance of container environments.

### Introduction

To understand security issues facing container technology we first have to understand what containers are. Containers are executable units of application that are packaged, along with their libraries and dependencies needed to run the application.

Containers allow developers to deploy and maintain applications seamlessly across different environments.

Different components of a container have different security risks associated with them. We are going to dive into images, registry, and Host Operating systems vulnerabilities.

### Key takeaways

1. What components of containers.

2. What are the vulnerabilities that are associated with the core components of containers.

3. How to mitigate these security risks.

### Reference

Microsoft Azure introduction to containers. https://azure.microsoft.com/en-us/overview/what-is-a-container/#beyond-containers

### PR DESCRIPTION
**Pull request by calebroHQ**
In this article, we have gone through what containers are, what are the security risks associated with the container technology, and how best to mitigate these risks.
      
This fixes #2216

### CODE REVIEW

**Comment by hectorkambow** [comment]
Thank you v much for your patience as we worked through our Repo re-structuring process. 

The purpose behind this restructure - was to better allow us to handle more volume and allow students to publish more articles and get more involved with the community (with open-source contribution and suggestions). 

With most of the changes coming to completion - we will be sure to get your article published asap. 

Thank you again - and we are looking forward to more amazing articles.

**Comment by calebroHQ** [comment]
Thank you, @hectorkambow

File: content/articles/fundamentals-of-container-security/index.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+### Container security best practices
+
+We've gone through several attack scenarios affecting container protection, but the risk of being breached can still be reduced by doing the following:
+
+1. Using container-specific operating systems- The attack surface is greatly reduced when operating systems designed to host containers are used as opposed to general-purpose operating systems. General-OS have many services and components, if one of this service has a security vulnerability the containers it is hosting can be compromised.
+
+2. Use of advanced malware analysis tool -With an advanced malware analysis tool, the engineers will be able to discover malware embedded in container images before it can cause harm to the organization.
+
+3. Proper storage of secret keys and tokens- Secrets should be stored outside of container images and provided dynamically at runtime as needed. Most orchestrators provide a secret key management feature.
```
Comment (srishilesh) [inline content/articles/fundamentals-of-container-security/index.md:55]:
```suggestion
3. Proper storage of secret keys and tokens - Secrets should be stored outside of container images and provided dynamically at runtime as needed. Most orchestrators provide a secret key management feature.
```

File: content/articles/fundamentals-of-container-security/index.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+4. Unbounded administrative access- Administrators should use an access management model in which users are only granted the ability to perform the specific task on the specific hosts, containers, and images their job roles require i.e the development team will only have access to containers used in development only.
```
Comment (srishilesh) [inline content/articles/fundamentals-of-container-security/index.md:57]:
```suggestion
4. Unbounded administrative access - Administrators should use an access management model in which users are only granted the ability to perform the specific task on the specific hosts, containers, and images their job roles require i.e., the development team will only have access to containers used in development only.
```

File: content/articles/fundamentals-of-container-security/index.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+### Introduction
+
+To grasp the security issues surrounding container technology, we must first understand what containers are and their role in the software development life-cycle.
+
+Containers are executable units of application that are packaged, along with their libraries and any dependencies required to run the application. Containers make it possible for developers to deploy and maintain applications seamlessly across different environments.
+
+Container technology has changed how developers and teams ship, deploy and maintain code, but this has come with its fair share of security risks. In this article, we will explore some of the risks, how they affect containers and how best counter-measures.
+
+
```
[the remark this answers is not in the document]
Comment (srishilesh) [reply content/articles/fundamentals-of-container-security/index.md:11]: Add links to other Section articles related to Docker and Container.

**Review by srishilesh — CHANGES_REQUESTED** [review]
Hey @calebroHQ,

Great article.

I've suggested a few changes. Please let me know when it's done, we can move on to the final review.

Thank you.

**Comment by calebroHQ** [comment]
@srishilesh, I made the changes as suggested.

**Comment by hectorkambow** [comment]
[Fundamentals of container security #2367.pdf](https://github.com/section-engineering-education/engineering-education/files/6563670/Fundamentals.of.container.security.2367.pdf)

Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept.
Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content.
Be sure to see our resources page to see more info on plagiarism and what is considered as such.

@calebroHQ

**Comment by srishilesh** [comment]
@calebroHQ 

The plagiarism score shows 26%

Please add appropriate references, and rephrase sentences to avoid higher plagiarism scores. Let me when it's done.

**Comment by calebroHQ** [comment]
hi @srishilesh, I have done the necessary revisions.

**Comment by srishilesh** [comment]
@hectorkambow 

Could you perform a plagiarism check again? Just to ensure if it's below the expected score.

Thank you.

**Comment by srishilesh** [comment]
@calebroHQ 

I did a quick plagiarism check by myself, and found it is 3%, which is accepted. Let me start the review tomorrow.

**Comment by calebroHQ** [comment]
@srishilesh, thank you

**Review by srishilesh — APPROVED** [review]
@calebroHQ 

Thanks for making the changes.

Moving for the final review.

Thank you.

**Comment by calebroHQ** [comment]
@srishilesh, thank you.

**Comment by hectorkambow** [comment]
good work gents @calebroHQ 

cc @srishilesh 

Please (if you both) could go over my final edits as well - as there were some sentences were a bit difficult to understand so I changed them - want to make sure I did not change the meaning of the sentences. Also see how we can avoid these errors for future articles 🚀

**Pull request merged**